internal/idle/idle.go

```diff
@@ -231,26 +240,29 @@ func (m *Manager) ExitIdleMode() error {
 		//   m.ExitIdleMode.
 		//
 		// In any case, there is nothing to do here.
-		return nil
+		return
 	}
 
-	// This can fail if resolver creation fails. In that case, we want to
-	// return the error to the caller so that the RPC can fail. But we still
-	// need to undo the idle entry process, and ensure that the idle timer is
-	// started again.
-	err := m.enforcer.ExitIdleMode()
+	m.enforcer.ExitIdleMode()
 
 	// Undo the idle entry process. This also respects any new RPC attempts.
 	atomic.AddInt32(&m.activeCallsCount, math.MaxInt32)
 	m.actuallyIdle = false
 
 	// Start a new timer to fire after the configured idle timeout.
 	m.resetIdleTimerLocked(m.timeout)
+}
 
-	if err != nil {
-		return fmt.Errorf("failed to exit idle mode: %v", err)
-	}
-	return nil
+// MarkAsExitedIdleMode instructs the Manager to update its internal state to
+// indicate that the channel has exited IDLE mode. This is only used by the gRPC
+// client when it exits IDLE mode manually from Dial.
+func (m *Manager) MarkAsExitedIdleMode() {
```

Comment:
This probably needs warnings on it.  Generally you can't use it in a situation where you can possibly race with anything else calling `ExitIdleMode` _or_ `OnCallBegin`, and you _must_ know you are really idle.

Bikeshedding:

`UnsafeSetNotIdle()`?  `UnsafeSetActive()`?

Aside: Do we already say "active" is the opposite of "idle" anywhere? Or are the only states "in idle mode" and "not in idle mode".  The latter feels a little awkward.  Maybe we can rename things to use "active" vs "idle":

```
ExitIdleMode() -> Activate()
EnterIdleModeForTesting() -> GoIdle()  // Do we really need to ForTesting this? It seems safe to use.
```